dali/kernels/normalize/normalize_gpu_impl_test.cu

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/normalize/normalize_gpu_impl.cuh"  // NOLINT
+#include <gtest/gtest.h>
+#include <cmath>
+#include <initializer_list>
+#include <random>
+#include <utility>
+#include "dali/kernels/kernel_manager.h"
+#include "dali/test/device_test.h"
+#include "dali/test/test_tensors.h"
+#include "dali/test/tensor_test_utils.h"
+
+namespace dali {
+namespace kernels {
+
+template <bool calc_inv_stddev, typename Out, typename In>
+void RefNormalize(
+    const OutTensorCPU<Out> &out,
+    const InTensorCPU<In> &in, const InTensorCPU<float> &base,
+    const InTensorCPU<float> &scale,
+    float epsilon, float global_scale, float shift,
+    TensorShape<> &data_pos, TensorShape<> &base_pos, TensorShape<> &scale_pos, int dim) {
+
+  int db = 0, ds = 0;
+  int64_t extent = 0;
+  if (dim < in.dim()) {
+    db = base.shape[dim] > 1 ? 1 : 0;
+    ds = scale.shape[dim] > 1 ? 1 : 0;
+    extent = in.shape[dim];
+  }
+  if (dim >= in.dim() - 1) {  // handles both last dimension and degenerate case
+    Out *optr = out(data_pos);
+    const In *iptr = in(data_pos);
+    const float *sptr = scale(scale_pos);
+    const float *bptr = base(base_pos);
+    for (int64_t i = 0, b = 0, s = 0; i < extent; i++, b += db, s += ds) {
+      float mul;
+      if (calc_inv_stddev) {
+        float x = sptr[s] * sptr[s] + epsilon;
+        mul = x ? rsqrt(x) * global_scale : 0;
+      } else {
+        mul = sptr[s] * global_scale;
+      }
+      optr[i] = ConvertSat<Out>(std::fma(iptr[i] - bptr[b], mul, shift));
+    }
+  } else {
+    for (int64_t i = 0, b = 0, s = 0; i < extent; i++, b += db, s += ds) {
+      data_pos[dim] = i;
+      base_pos[dim] = b;
+      scale_pos[dim] = s;
+      RefNormalize<calc_inv_stddev>(out, in, base, scale, epsilon, global_scale, shift,
+                                    data_pos, base_pos, scale_pos, dim + 1);
+    }
+  }
+}
+
+
+/**
+ * @brief Reference normalization of a single tensor
+ *
+ * If base/scale has an extent of 1 in any given dimension, it's broadcast along this axis.
+ *
+ * @param calc_inv_stddev if true, `scale` is assumed to contain standard deviation, which
+ *                        is subsequently regularized using given epsilon value
+ */
+template <typename Out, typename In>
+void RefNormalize(
```

Comment:
What value `template <bool calc_inv_stddev, typename Out, typename In>
void RefNormalize` adds comparedto `template <typename Out, typename In>
void RefNormalize(` ?

Reply:
There's an `if` in the inner loop - and I like it being a compile-time one.